### PR TITLE
Support PD disaggregation

### PR DESCRIPTION
**Pull request by zhenwei-intel**
Support PD disaggregation:
- [x]  Non-DeepSeek Models​ are now supportted
- [x]  DeepSeek Models​:  Done (migrate from https://github.com/HabanaAI/vllm-fork/pull/1050)

How to test: pls follow `pd_xpyd/readme.md`

llama3-8B:
acc: 0.7453(same as w/o PD disagg)
image

### CODE REVIEW

**Comment by jikunshang** [comment]
can you rebase and fix format related issue first?

**Review by jikunshang — COMMENTED** [review]
do we need all of these scripts? I think they are for test purpose.

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 1]
```diff
@@ -34,7 +41,19 @@ def __init__(
...
+        self.k_v_head_size = self.k_head_size + self.v_head_size
+        self.block_size = 128
+        max_num_blocks = 1000
+        self.block_indice_place_holder = torch.zeros(max_num_blocks,
+                                                     dtype=torch.int,
+                                                     device="hpu")
+        self.padded_length_tensor = torch.zeros(1,
+                                                dtype=torch.int,
+                                                device="hpu")
+        self.is_deepseek = config.model_config.is_deepseek_mla
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:56]: I don't think we should use deepseek, we should use is MLA instead.

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 2]
```diff
@@ -34,7 +41,19 @@ def __init__(
         self.config = config.kv_transfer_config
         self.kv_helper = kv_helper(config)
         self.local_tp_rank = local_rank
-
+        self.rank = rank
+        self.k_head_size = 64
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:45]: Please unify `k_head_size` `v_head_size` `kv_head_size` and `num_heads, head_size = self.kv_helper.get_model_args(model_executable)`

File: pd_xpyd/readme.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+```bash
+apt install etcd -y
+pip3 install mooncake-transfer-engine==0.3.0
+```
+
+### Install from Source
+
+```bash
+# Install required packages
+apt install git wget curl net-tools sudo iputils-ping etcd -y
```
Comment (jiafuzha) [inline pd_xpyd/readme.md:18]: etcd is duplicated

File: pd_xpyd/readme.md [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+# Clone the Mooncake repository
+git clone https://github.com/kvcache-ai/Mooncake.git -b v0.3.0
```
Comment (jiafuzha) [inline pd_xpyd/readme.md:21]: It's duplicated with above mooncake-transfer-engine too.

Comment (jikunshang) [reply pd_xpyd/readme.md:21]: we should keep both pip install and manual build/install step for user.

File: pd_xpyd/readme.md (hunk 3 above)
Comment (zhenwei-intel) [reply pd_xpyd/readme.md:18]: Here are two installation methods, and you only need to choose one.

File: pd_xpyd/readme.md [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+> **Note**: This document does not cover MLA data parallel setup.
+
+## Mooncake Installation
+
+### Install via pip
+
+```bash
+apt install etcd -y
+pip3 install mooncake-transfer-engine==0.3.0
```
[the remark this answers is not in the document]
Comment (zhenwei-intel) [reply pd_xpyd/readme.md:11]: I didn't find any issues, but pinning to beta would be safer.

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py (hunk 1 above)
Comment (zhenwei-intel) [reply vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py]: thanks, update to is_mla

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py (hunk 2 above)
Comment (zhenwei-intel) [reply vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py]: unified, thanks.

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 6]
```diff
@@ -34,7 +41,15 @@ def __init__(
         self.config = config.kv_transfer_config
         self.kv_helper = kv_helper(config)
         self.local_tp_rank = local_rank
-
+        self.rank = rank
+        self.block_size = 128
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:45]: self.block_size = vllm_config.cache_config.block_size

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 7]
```diff
@@ -34,7 +41,15 @@ def __init__(
...
+        max_num_blocks = 1000
+        self.block_indice_place_holder = torch.zeros(max_num_blocks,
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:47]: remove if not in use?

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 8]
```diff
@@ -192,6 +213,223 @@ def recv_kv_caches_and_hidden_states(
...
+            current_slot_mapping = model_input.attn_metadata.slot_mapping[
+                idx][:padded_total_size]
+            self.padded_length_tensor[0] = padded_total_size
+            htorch.core.mark_step()
+            # ==== graph should start here ======
+            keys, values = [], []
+            for layer_id in range(start_layer, end_layer):
+                kv_cache = kv_caches[layer_id - start_layer]
+                if self.kv_helper.is_mla():
+                    key_cache = kv_cache[0].reshape(-1, num_heads, head_size)
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:255]: Please leave a comments why we only use key_cache.

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 9]
```diff
@@ -192,6 +213,223 @@ def recv_kv_caches_and_hidden_states(
...
+        # In this case we need to do prefilling to recompute missing KV cache
+        # and hidden states.
+        bypass_model_exec = True
+        input_tokens_tensor_cpu = model_input.input_tokens.to("cpu")
+        torch.hpu.synchronize()
+
+        seq_lens = model_input.attn_metadata.seq_lens_tensor.tolist()
+        block_indices_list = attn_metadata.block_indices.tolist()
+        hidden_or_intermediate_states_for_one_req: list[torch.Tensor] = []
+        input_tokens_list = []
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:301]: not used?

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 10]
```diff
@@ -192,6 +213,223 @@ def recv_kv_caches_and_hidden_states(
...
+        num_computed_tokens_list = []
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:302]: same as above?

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 11]
```diff
@@ -192,6 +213,223 @@ def recv_kv_caches_and_hidden_states(
...
+        start_block_idx = 0
+        num_heads, head_size = self.kv_helper.get_model_args(model_executable)
+
+        for idx, slen in enumerate(seq_lens):
+            current_tokens = input_tokens_tensor_cpu[idx][:slen]
+            num_blocks = (slen + 127) // 128
+            end_block_idx = start_block_idx + num_blocks
+            block_indices_tensor = torch.tensor(
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:310]: this may cause graph recompile I guess.

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 12]
```diff
@@ -192,6 +213,223 @@ def recv_kv_caches_and_hidden_states(
...
+                    if self.kv_helper.is_mla():
+                        padding_k_tensor = torch.zeros(
+                            (self.block_size, head_size),
+                            dtype=self.dtype,
+                            device="hpu")
+                        self.cache_k(
+                            padding_k_tensor.unsqueeze(0),
+                            key_cache,
+                            attn_metadata.
+                            block_indices[start_block_idx:end_block_idx],
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:333]: where define this?

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 13]
```diff
@@ -192,6 +213,223 @@ def recv_kv_caches_and_hidden_states(
...
+            # get roi for current seq
+            load_key_prefix = self.tensor_hash(current_tokens)
+            # For deepseek, we only need recv first rank
+            load_kvcache_key = f"{load_key_prefix}_0"
+            remote_kv = self.kv_store.get(load_kvcache_key)
+            hidden_key = f"{load_key_prefix}_hidden_0"
+            hidden = self.kv_store.get(hidden_key)
+
+            if remote_kv is None or hidden is None:
+                logger.warning("Didn't find any match, load_key_prefix: %s",
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:375]: I suggest to add rank id in log.

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py [hunk 14]
```diff
@@ -192,6 +213,223 @@ def recv_kv_caches_and_hidden_states(
...
+            start_block_idx = end_block_idx
+            htorch.core.mark_step()
+
+        if not bypass_model_exec:
+            # Some of the KV cache is not retrieved
+            # Here we will fall back to normal model forwarding
+            # But optionally you can adjust model_input so that you only do
+            # prefilling on those tokens that are missing KV caches.
+            logger.warning(
+                "[rank%d]: Failed to receive all KVs and hidden "
```
Comment (jikunshang) [inline vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:421]: use self.rank and self.local_rank instead?

File: vllm/worker/hpu_model_runner.py [hunk 15]
```diff
@@ -2582,6 +2595,57 @@ def make_model_input_from_broadcasted_tensor_dict(
...
+            kv_caches[0][0].numel() == 0)
+        # check if the current run is prefill
+        return self.vllm_config.kv_transfer_config.is_kv_consumer and (
+            not is_profile_run) and is_prefill_run
+
+    def need_send_kv(self, model_input, kv_caches, warmup_mode) -> bool:
+        """Check if we need to send kv-cache to the other worker.
+        We need to send KV when
+            1. current vLLM instance is KV cache producer/prefill vLLM instance
+            2. this batch is not a profiling run
```
Comment (jikunshang) [inline vllm/worker/hpu_model_runner.py:2627]: or a warmup run.

File: vllm/worker/hpu_model_runner.py [hunk 16]
```diff
@@ -2801,18 +2865,69 @@ def try_revert_dummy_output_tokens():
...
+                            selected_token_indices)
+                        if warmup_mode:
+                            torch.hpu.synchronize()
+                            import torch.distributed as dist
+                            if dist.is_initialized():
+                                dist.barrier()
+                else:
+                    logger.debug("Bypassing model execution")
+
+                # torch.hpu.synchronize()
```
Comment (jikunshang) [inline vllm/worker/hpu_model_runner.py:2914]: can remove this

File: vllm/worker/hpu_model_runner.py [hunk 17]
```diff
@@ -2801,18 +2865,69 @@ def try_revert_dummy_output_tokens():
...
+                        # model_executable is used to know which layer the
+                        # current worker is working on, so that we can send KV
+                        # for only those layers.
+                        self.get_model(),
+                        model_input,
+                        kv_caches,
+                        hidden_states,
+                    )
+                    now = time.time()
+                    logger.info("KV transfer send time: %f", now - cur_time)
```
Comment (jikunshang) [inline vllm/worker/hpu_model_runner.py:2930]: logger.debug

File: vllm/worker/hpu_model_runner.py [hunk 18]
```diff
@@ -2801,18 +2865,69 @@ def try_revert_dummy_output_tokens():
...
+                            torch.hpu.synchronize()
+                            import torch.distributed as dist
+                            if dist.is_initialized():
+                                dist.barrier()
+                else:
+                    logger.debug("Bypassing model execution")
+
+                # torch.hpu.synchronize()
+                # Sending KV cache in distributed KV cache transfer setting
+                # NOTE: the send operation is non-blocking
```
Comment (jikunshang) [inline vllm/worker/hpu_model_runner.py:2916]: it's blocking right?

File: vllm/worker/hpu_model_runner.py [hunk 19]
```diff
@@ -2801,18 +2865,69 @@ def try_revert_dummy_output_tokens():
...
+                cur_time = time.time()
```
Comment (jikunshang) [inline vllm/worker/hpu_model_runner.py:2917]: it's not suggested I suppose.

File: vllm/worker/hpu_worker.py [hunk 20]
```diff
@@ -513,12 +514,13 @@ def get_cache_block_size_bytes(self) -> int:
...
 
 def init_worker_distributed_environment(
-    parallel_config: ParallelConfig,
+    vllm_config: VllmConfig,
     rank: int,
     distributed_init_method: Optional[str] = None,
     local_rank: int = -1,
 ) -> None:
     """Initialize the distributed environment."""
+    parallel_config = vllm_config.parallel_config
```
Comment (jikunshang) [inline vllm/worker/hpu_worker.py:523]: we can change this part first in a seprate PR first.

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py (hunk 6 above)
Comment (zhenwei-intel) [reply vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:45]: updated, thx

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py (hunk 7 above)
Comment (zhenwei-intel) [reply vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py]: updated, thx

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py (hunk 10 above)
Comment (zhenwei-intel) [reply vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py]: updated, thx

File: vllm/worker/hpu_model_runner.py (hunk 16 above)
Comment (zhenwei-intel) [reply vllm/worker/hpu_model_runner.py]: updated, thx

File: vllm/worker/hpu_model_runner.py (hunk 18 above)
Comment (zhenwei-intel) [reply vllm/worker/hpu_model_runner.py]: updated, thx

File: vllm/worker/hpu_model_runner.py (hunk 19 above)
Comment (zhenwei-intel) [reply vllm/worker/hpu_model_runner.py]: remove this log info

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py (hunk 8 above)
Comment (zhenwei-intel) [reply vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:247]: could you help explain this?

Comment (jikunshang) [reply vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:247]: please point to `HPUMLAAttentionBackend::get_kv_cache_shape` method.

File: vllm/worker/hpu_worker.py (hunk 20 above)
Comment (zhenwei-intel) [reply vllm/worker/hpu_worker.py:523]: will keep it in this PR

File: vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py (hunk 8 above)
Comment (zhenwei-intel) [reply vllm/distributed/kv_transfer/kv_connector/mooncake_store_connector.py:249]: Thanks

**Comment by zhenwei-intel** [comment]
> is this PR still needed? Are You waiting for review or is it not needed anymore?

Still waiting for review, this PR is to be merged.

File: vllm/worker/hpu_worker.py [hunk 21]
```diff
@@ -513,12 +514,13 @@ def get_cache_block_size_bytes(self) -> int:
 
 
 def init_worker_distributed_environment(
-    parallel_config: ParallelConfig,
+    vllm_config: VllmConfig,
```
Comment (xuechendi) [inline vllm/worker/hpu_worker.py:517]: This change is to align with GPU arguments, right?
Based on the codes, I don't see necessity for this change

File: vllm/distributed/kv_transfer/kv_lookup_buffer/mooncake_store.py [hunk 22]
```diff
@@ -154,7 +158,7 @@ def _get_impl(
             device_id_tensor = loaded_tensors["device_id"]
             device_id = int(device_id_tensor.item())
             device = torch.device(
-                'cuda', device_id) if device_id >= 0 else torch.device('cpu')
+                'hpu', device_id) if device_id >= 0 else torch.device('cpu')
```
Comment (xuechendi) [inline vllm/distributed/kv_transfer/kv_lookup_buffer/mooncake_store.py:161]: Can we make it a variable based on platform. Changing 'cuda' to 'hpu' might be tricky for upstream rebase.

File: vllm/distributed/kv_transfer/kv_lookup_buffer/mooncake_store.py [hunk 23]
```diff
@@ -125,8 +125,12 @@ def _put_impl(
         value: torch.Tensor,
     ) -> None:
         """Put KVCache to Mooncake Store"""
-        device_id = value.device.index if value.device.type == 'cuda' else -1
-        device_tensor = torch.tensor(device_id, dtype=torch.int32)
+        device_id = value.device.index if value.device.type == 'hpu' else -1
```
[the remark this answers is not in the document]
Comment (zhenwei-intel) [reply vllm/distributed/kv_transfer/kv_lookup_buffer/mooncake_store.py]: thanks, update it.

File: pd_xpyd/readme.md [hunk 24]
```diff
@@ -0,0 +1,167 @@
...
+
+# Check RDMA devices and network interfaces
+ibdev2netdev
+# Example output:
+# mlx5_0 port 1 ==> ens108np0 (Up)
+# mlx5_1 port 1 ==> ens9f0np0 (Up)
+# ...
+```
+
+> **Note**: If you are running RDMA inside a Docker container, add the `--privileged` flag to the `docker run` command to ensure proper hardware and network access.
```
Comment (jiafuzha) [inline pd_xpyd/readme.md:54]: Please also mention enabling Transparent Huge Page here. Otherwise, we'll see decoding degradation compared to TCP.

**Review by xuechendi — APPROVED** [review]
